# Having Some Fun with Project Euler Problems

[Project Euler](http://projecteuler.net) is a collection of mathematical and computational problems that serves as an  extremely useful resource to test and improve your both your problem solving and coding abilities. The site is composed of different mathematical problems and tests your ability to come up with a solution using the programming language of your choice. I absolutely love the site. Here are just a few of the relatively easier problems. 

It's important to note that Project Euler isn't so much concerned about how long it takes you to develop your algorithm, or even how elegant it is. Each problem is designed according to the "one-minute rule," meaning that regardless of your algorithm, you should be able to solve the problem on a modestly powerful computer within one minute or less. 

With that in mind, let's go through some of easier problems. 

## 1. Multiples of 3 and 5: Find the sum of all multiples of 3 and 5 below 1,000. 

This problem is fairly straightforward. A multiple of a number $n$ is any number that can be evenly divided by $n$. Thus, if $n = 3$ , multiples of $3$ include $\{3,6,12,15,18, \ldots \}$. Here is a simple solution to finding the sum of all multiples of 3 and 5 below 1,000.

In [12]:
multiples = []
for n in range(1,1000):
    if (n % 3 == 0) or (n % 5 == 0):
        multiples.append(n)

sum(multiples)

233168

We can even shorten the above code using list comprehension, thereby reducing 4 lines of code into 1. 

In [13]:
multiples = [n for n in range(1,1000) if (n % 3 == 0) or (n % 5) == 0]

sum(multiples)

233168

## 2. Even Fibonacci Numbers: Find the sum of even-valued terms in a Fibonacci sequence whose value does not exceed 4 million. 

If you're unfamiliar with the Fibonacci sequence, it is a sequence of numbers in which the next number in the series can be calculated by the preceding two numbers. The following is a Fibonacci sequence $\{0,1,1,2,3,5,8,13,21,34,55,89,...\}$. Note that the $0$ is optional in the sequence and doesn't necessarily have to be included. That is, you don't necessarily have to start a Fibonacci sequence with $0$. 

A very popular way of computing the $n^{th}$ term of a Fibonacci sequence is by creating a **recursive function** such as the following:

In [14]:
def fib(n):
    if n <= 2:
        return 1
    else:
        return fib(n-1) + fib(n-2)

We can test out our function to find the first 10 terms in a Fibonacci sequence.

In [15]:
for n in range(1,10):
    print fib(n)

1
1
2
3
5
8
13
21
34


However, there is one major problem with this approach. The function that we created works well up until a certain index in the sequence before it eventually reaches a point in which it becomes computationally slower to calculate the $n^{th}$ number in the sequence. That being said, I will not venture to calculate $\text{fib}(100)$ to illustrate my point, as it will take quite a long time. 

Fortunately, there is a more computationally efficient solution that not only gives us the $n^{th}$ number in the Fibonacci sequence, but by creating a function that takes in a parameter $n$ where $n$ represents how many numbers in the sequence you wish to return, we can solve the problem much more efficiently. Let's see how we might create such a function.

In [16]:
def fib2(n):
    fibs = [0, 1]
    for n in range(2, n+1):
        fibs.append(fibs[-1] + fibs[-2])
    return fibs

Let's take a look at how the function works. It starts off by creating a list of the first two terms of the Fibonacci sequence. From there, we create a for loop ranging from $ 2 $ to $n+1$ and append the results using yet another recursive function that sums the previous two numbers in the sequence. It then returns a Fibonacci sequence. Let's test our function. 

In [17]:
fib2(20)  # returns the first 20 terms of the Fibonacci sequence.

[0,
 1,
 1,
 2,
 3,
 5,
 8,
 13,
 21,
 34,
 55,
 89,
 144,
 233,
 377,
 610,
 987,
 1597,
 2584,
 4181,
 6765]

Let's see how well this function does in computing the first 100 terms. 

In [18]:
fib2(100)

[0,
 1,
 1,
 2,
 3,
 5,
 8,
 13,
 21,
 34,
 55,
 89,
 144,
 233,
 377,
 610,
 987,
 1597,
 2584,
 4181,
 6765,
 10946,
 17711,
 28657,
 46368,
 75025,
 121393,
 196418,
 317811,
 514229,
 832040,
 1346269,
 2178309,
 3524578,
 5702887,
 9227465,
 14930352,
 24157817,
 39088169,
 63245986,
 102334155,
 165580141,
 267914296,
 433494437,
 701408733,
 1134903170,
 1836311903,
 2971215073,
 4807526976,
 7778742049,
 12586269025,
 20365011074,
 32951280099,
 53316291173,
 86267571272,
 139583862445,
 225851433717,
 365435296162,
 591286729879,
 956722026041,
 1548008755920,
 2504730781961,
 4052739537881,
 6557470319842,
 10610209857723,
 17167680177565,
 27777890035288,
 44945570212853,
 72723460248141,
 117669030460994,
 190392490709135,
 308061521170129,
 498454011879264,
 806515533049393,
 1304969544928657,
 2111485077978050,
 3416454622906707,
 5527939700884757,
 8944394323791464,
 14472334024676221,
 23416728348467685,
 37889062373143906,
 61305790721611591,
 99194853094755497,
 16050

Relatively much faster than our first function that we defind. **I would NOT advise not to calculate the 100th term of the Fibonacci sequence using our first fib function. You will find yourself waiting for an eternity**.

So, now that we have a function that is computationaly efficient, let's solve the problem of finding the sum of the even-valued terms in a Fibonacci sequence whose value does not exceed 4,000,000. 

In [19]:
# I do not know at precisely what index is a number in a Fibonacci sequence is less than 4M, and I'm sure there's way
# of easily calculating that, but for now, let's just do some trial and error.

sequence = fib2(50)
sequence

[0,
 1,
 1,
 2,
 3,
 5,
 8,
 13,
 21,
 34,
 55,
 89,
 144,
 233,
 377,
 610,
 987,
 1597,
 2584,
 4181,
 6765,
 10946,
 17711,
 28657,
 46368,
 75025,
 121393,
 196418,
 317811,
 514229,
 832040,
 1346269,
 2178309,
 3524578,
 5702887,
 9227465,
 14930352,
 24157817,
 39088169,
 63245986,
 102334155,
 165580141,
 267914296,
 433494437,
 701408733,
 1134903170,
 1836311903,
 2971215073,
 4807526976,
 7778742049,
 12586269025]

It definitely looks like that there is a value that is less than 4,000,000 within the first 50 terms of the Fibonacci sequence. We can manually check it, but since this is an exercise in programming, why don't we use code to check where the value is less than 4M. 

In [20]:
for n in sequence:
    if n < 4000000:
        print sequence.index(n)

0
1
1
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33


The 33rd number in the Fibonacci sequence is the number that is less than 4M. This is enough information to solve our problem. 

In [21]:
fib_series = fib2(33)
even_fibs = []

for n in fib_series:
    if n % 2 == 0:
        even_fibs.append(n)
        
even_fibs

[0, 2, 8, 34, 144, 610, 2584, 10946, 46368, 196418, 832040, 3524578]

Now that we have a list of all the even-valued numbers in the sequence, we can just simply take the sum.

In [22]:
sum(even_fibs)

4613732

And voilà. There you have it. The sum of all even numbers in a Fibonacci sequence whose value does not exceed 4,000,000. 

## 3. What is the smallest number that can be evenly divided by all numbers from 1 to 20? 

In this problem, we're essentially asked to find the least common multiple (LCM) for a set of all numbers. If you need a refresher, the least common multiple between two numbers is the smallest number (greater than 0) that is a multiple of both numbers. That is, given two numbers $a$ and $b$, the least common multiple of $a$ and $b$ is the smallest number that can be evenly divided by both $a$ and $b$. For example, the least common multiple of 6 and 4 is 12, since $\frac{12}{6}=2$ and $\frac{12}{4}=3$. 

Finding the LCM of two numbers is quite simple (depending on the two numbers, of course). However, it becomes a little trickier, to say the least, to find the least common multiple for 3 or more numbers. For example, what is the least common multiple of all numbers from 1 to 10? Well, they give you the answer in this question, and the answer is 2,520. We can even do a quick check using very inefficient code. 

In [23]:
i = 10  # Since we're looking for the smallest number that is evenly divisible by 1 through 10, it makes sense to start
        # with a number greater than 10, as any number less than 10 will reslut in at least one fraction. 
remainders = [i % n for n in range(1, 11)]

for n in remainders:
    while sum(remainders) != 0:
        i += 1
        remainders = [i % n for n in range(1, 11)]
        sum(remainders)

Let's examine the code above. I initially created a list in which I store the value of the remainder of i % n, where n is an integer from 1 to 10. Then, I created a for loop that continues to calculte increments i by 1 and calculate i % n for all n in remainders until the sum of remainders is 0 to find the LCM. The logic behind this is that if all numbers in the remainders list is divisible by i, then the remainder will be 0. Thus, if the remainders list contains all 0s, we find our LCM. Let's see if we get 2,520 using our inefficient code. 

In [24]:
i

2520

It checks out. The smallest number that is divisible by all numbers 1 through 10 is 2,520. The problem with this code, and why it is inefficient, is that it works fine for the given range (1 through 11 not inclusive). However, the problem asks for the smallest number that is evenly divisible by 1 through 20, which is a greater range. Using this code will not work, as it is computationally slow. However, there's a simple solution to this that is much more efficient. 

It involves using Euclid's algorithm for finding the greatest common denominator and proceeding to create two functions to find the LCM between two or more numbers. If you're unfamiliar with Euclid's algorithm for finding the greatest common denominator (GCD), the GCD of two numbers A and B, 

    1. If A = 0, then GCD(A, B) = B. 
    2. If B = 0, then GCD(A, B) = A. 
    3. If neither A or B is 0, then express A in quotient remainder form (A = B*Q + R)
    4. Find GCD(B, R) using Euclid's algorithm until B = 0.
    
Let's create our functions. 

In [25]:
def gcd(a, b):
    """Returns GCD of a and b"""
    while b:
        a, b = b, a % b 
    return a
    
def lcm(a, b):
    """Returns LCM of a and b"""
    return a * b//gcd(a, b)

def lcm_multi(*args):
    """Returns LCM of 2 or more numbers"""
    return reduce(lcm, args)

We now have everything we need to compute the LCM of all integers from 1 through 20. 

In [27]:
lcm_multi(*range(1, 20))

232792560

There you have it. The smallest number that can be divided by 1 through 20 is 232,792,560.

## 4. What is the largest prime factor of 600,851,475,143? 

Let's go back to elementary school and revisit what prime numbers are. A **prime** number is any integer $n>1$ whose only factors are itself and 1. Thus, 2, 5, and 7 are all prime numbers since for each of those numbers, the only factors are 2 and 1, 5 and 1, and 7 and 1. A **composite** number is the opposite of a prime number in that it is a number that has factors beyond itself and 1. That it is, it can be evenly divided by another number besides 1 and itself. An example would be 10 and 36. 10 has factors of 1, 2, 5, and 10. 36 has factors of 1, 2, 2, 3, 3, and 6. Multiplying all prime factors will result in the original number. In the case of 36, for example, $2 \times 2 \times 3 \times 3 = 4 \times 9 = 36$. 

Don't let the large size of the number scare you. The solution to this problem is quite simple. We can define a function that finds all prime factors for any given integer $n>1$. 

In [28]:
def prime_factors(n):
    if n <= 1:
        pass
    else:
        i = 2
        factors = []
        while i**2 <= n:
            if n % i:
                i += 1
            else:
                n //= i
                factors.append(i)
        if n > 1:
            factors.append(n)
    return factors

Let's check if our code works. 

In [29]:
prime_factors(36)

[2, 2, 3, 3]

In [30]:
prime_factors(81)

[3, 3, 3, 3]

Sure enough, it does. Now, I've already gone ahead and tried passing 600,851,475,143 to the function, and it is well within the "one-minute" rule. So, let's try it. 

In [33]:
prime_factors(600851475143)

[71, 839, 1471, 6857]

There we have it. The largest prime factor of 600,851,475,143 is 6857. However, let's just check to see if 6857 is really prime. By passing it through our function, it should return 6857 if it is in fact prime. 

In [34]:
prime_factors(6857)

[6857]

Checks out. 6857 is a prime number. 

## 5. Find the largest palindrome made from the product of two 3-digit numbers.

A palendrome is a number that reads the same backwards as forward. For example, 101, 2002, 181, 9669 are all palendromic numbers. The solution:

In [37]:
a = []
b = []
a.extend(range(100,1000))
b.extend(range(100,1000))
palindromes = []

for i in a:
    for j in b:
        i * j
        if str(i * j) == str(i * j)[::-1]:
            palindromes.append(i * j)

Let's first make sure our list actually consists of palendromic numbers. 

In [39]:
palindromes[:10]

[10201, 11211, 12221, 13231, 14241, 15251, 16261, 17271, 18281, 19291]

Looks like it does. Now let's see what the largest palindromic number between two 3-digit numbers is. 

In [40]:
max(palindromes)

906609

The largest palendromic number between two 3-digit numbers is 906609. 